pgrx-examples/hooks/src/lib.rs

```diff
@@ -0,0 +1,225 @@
+use pgrx::prelude::*;
+
+::pgrx::pg_module_magic!(name, version);
+
+fn delete_must_have_a_where(query: PgBox<pg_sys::Query>) {
+    if query.commandType != pg_sys::CmdType::CMD_DELETE {
+        return ();
+    }
+
+    // SAFETY: the jointree is always defined in a DELETE Query
+    let jointree = unsafe { PgBox::from_pg(query.jointree) };
+
+    if jointree.quals.is_null() {
+        panic!("DELETE queries must have a WHERE clause");
+    }
+}
+
+fn only_superusers_can_truncate(pstmt: PgBox<pg_sys::PlannedStmt>) {
+    if !unsafe { pgrx::is_a(pstmt.utilityStmt, pg_sys::NodeTag::T_TruncateStmt) } {
+        return ();
+    }
+
+    if !unsafe { pg_sys::superuser() } {
+        panic!("Only superusers can truncate")
+    }
+}
+
+unsafe fn register_hooks() {
+    //
+    // Post Parse Analyze hook
+    //
+    static mut PREV_POST_PARSE_ANALYZE_HOOK: pg_sys::post_parse_analyze_hook_type = None;
+    PREV_POST_PARSE_ANALYZE_HOOK = pg_sys::post_parse_analyze_hook;
+    pg_sys::post_parse_analyze_hook = Some(post_parse_analyze_hook);
+
+    // The hook functions signatures may change between major version
+    // For instance: in the post_parse_analyze hook, the JumbleState struct
+    // appeared in Postgres 14
+    // In that case, we need some conditional compilation to declare the
+    // proper signature for each version
+    #[cfg(feature = "pg13")]
+    #[pg_guard]
+    unsafe extern "C-unwind" fn post_parse_analyze_hook(
+        parse_state: *mut pg_sys::ParseState,
+        query: *mut pg_sys::Query,
+    ) {
+        delete_must_have_a_where(PgBox::from_pg(query));
+        if let Some(prev_hook) = PREV_POST_PARSE_ANALYZE_HOOK {
+            prev_hook(parse_state, query);
```

Comment:
```suggestion
            pgrx::pg_sys::ffi::pg_guard_ffi_boundary(|| prev_hook(parse_state, query));
```

The other three calls to `prev_hooks` also need to be modified in the same way.

Reply:
👋🏻 (I'm new here)

The documentation on `pg_guard_ffi_boundary` is all internal. When/Where is it appropriate to use? When/Where is it _necessary_ to use?

Reply:
> When/Where is it _necessary_ to use?

It's **necessary** to use if you're calling any PostgreSQL-exported functions or extension-exported functions.

In most cases, you do not need to explicitly use this function because the functions in `pgrx::pg_sys` are rewritten through macros as follows:

```rust
#[pgrx::pg_guard]
extern "C" {
    fn psql_fn();
}
// The code expands to:
/************************
fn psql_fn() {
    extern "C" {
        fn psql_fn();
    }
    pgrx::pg_sys::ffi::pg_guard_ffi_boundary(|| psql_fn())
}
************************/
```

However, for any function pointers coming from PostgreSQL, we cannot rewrite them in this way. Therefore, you need to manually call `pg_guard_ffi_boundary`.

```rust
#[pgrx::pg_guard]
#[unsafe(no_mangle)]
extern "C-unwind" fn this_fn_can_be_called_by_postgresql(p: extern "C-unwind" fn()) {
    // `p` comes from PostgreSQL, so calling `p()` is dangerous!
    // let's use `pg_guard_ffi_boundary`
    pgrx::pg_sys::ffi::pg_guard_ffi_boundary(|| p())
}
```

As for why we need to wrap all PostgreSQL-exported function-call with `pg_guard_ffi_boundary`, please refer to https://github.com/pgcentralfoundation/pgrx/blob/develop/pgrx-pg-sys/src/submodules/ffi.rs.

Reply:
Imho this is the most impressive feature of PRGX.